### DISCUSSION TITLE
Closure syntax and implementation

### DISCUSSION THREAD

**Discussion by olup** [vlang/v#11763] [Ideas]
This issue just to lay down proposals for `closure` syntax and implementation, as discussed in the `#syntax` thread on Discord.

A closure is a function that retains some variables present in its definition's environment on top of the usual arguments that one can pass at calling time.

Similar to PHP or Rust, and as opposed to javascript or Go, it was proposed to require explicitly the "inherited" values. Types don't need to be indicated as those values have already been typed in the parent environment and can be inferred.

From what I gathered in the thread we have two main options : 

Use a new keyword. `use` was proposed, or `using`. That keyword could stand right after the argments definition:

`closure := fn (foo string) use (some_value_name) string { ... }`

Or use some syntax elements. Brackets where cited or - similarly to Rust - the pipe symbol:

`closure := fn (foo string) |some_value_name| string {...}`

I'll add comments in the issue so that we can +1 on them - or add other proposals as well.

**Comment by olup** [comment]
`closure := fn (foo string) use (some_value_name) string { ... }`

**Comment by olup** [comment]
`closure := fn (foo string) |some_value_name| string {...}`

**Comment by JalonSolov** [comment]
Figured out why the `|x|` syntax bothers me.  In Math, `|x|` means `absolute value of x`, and that's not what it means here.

I'm not a huge fan of 'use' keyword, either, but can't think of anything better atm.

**Comment by esquerbatua** [comment]
As JalonSolov says more or less, I don't like add more keywords like `use`, by that I like an little bit the use of `|x|` because it's more readable but it's significant it's "absolute of".

Maybe using another symbol? like `!x!` or '#x#', but it doesn't convince me either to 100%

**Comment by olup** [comment]
On the plus side the pipe symbol means "pipe", "transfer to", as in the pipe Linux shell syntax - which works with the use case. We virtually "pipe" the values in the closure.

**Comment by JalonSolov** [comment]
With a single pipe, yes.  Not one on each side.

**Comment by ntrel** [comment]
Why declare which variables the closure will capture? Just capture any variables that are used.

**Comment by olup** [comment]
Yes, I am not sure why the "use"  declaration was proposed in the fist
place. To make the compiler easier to write ? To mark the closure as well ?
To make it more explicit to the coder ?

Go and Rust and Javascript actually don't require a capture syntax, so why
should we ? However, if no capture syntax is used, we have to come up with
something to signify a closure in the code (or make ALL anonymous functions
closures ?)

**Comment by JalonSolov** [comment]
I believe the idea for declaring which variables are captured is to make the code safer.  Otherwise you can't be sure (without carefully reading the code) which variables might get modified within the closure.  A simple typo could wreak havoc on the rest of the program.

**Comment by olup** [comment]
Closure cannot modify their inherited variables in the parent scope.

Le lun. 11 janv. 2021 à 14:45, JalonSolov <notifications@github.com> a
écrit :

**Comment by JalonSolov** [comment]
Well, actually you can, with mutable variables and `unsafe` blocks.  :-\

**Comment by ntrel** [comment]
> you can, with mutable variables and unsafe blocks

Why use unsafe to do that when you could just use a mutable parameter to the closure?

> A simple typo could wreak havoc on the rest of the program.

(As it can everywhere outside the closure)

**Comment by JalonSolov** [comment]
Because it's there, _someone_ will use it (my biggest gripe against having `unsafe` at all).

At the least, if you know exactly which variables the closure is allowed to touch, you know it WASN'T the closure code if something else gets modified.  Unless `unsafe` is used, and they do funky things with pointers...

**Comment by olup** [comment]
> Well, actually you can, with mutable variables and `unsafe` blocks. :-\

Sorry, but I don't see what you are referring to. There is no proper closure implementation yet in V, but I think the proposal should be that you cannot mutate the parent context. If possible, use of an identifier outside of its context should be forbidden.

The problem with unsafe is real, but does not seem related to closure implementation per se, and I am not sure declaring the imported variables will really help with that.

However, it might be clearer to read, or easyer to parse. But most languages I know doesn't need that and I never felt any dificulties knowing what came form the context.

**Comment by kitech** [comment]
It much easy to use without explict capture variables.
Avoid stupid operation is good, but not easy to prevent that. even in rust or java or go can do some magic in `unsafe` block if someone want.

**Comment by olup** [comment]
If capturing syntax is to be avoided, maybe we should offer here a syntax
proposal to diferenciate closure from simple anonymous functions (I think
rust has that) ? Or again, should every anonymous function be a closure (Go
and JS works like that) ?

**Comment by Ben-Fields** [comment]
Explicit capture is a must as I see it, for a similar reason that functions are pure by default.

```v
closure := fn (foo string) string | val_a, val_b { /*...*/ }
```
Or something to split up the header, either syntax or a vfmt solution.
```v
closure := fn (foo string) string { 
    keep val_a, val_b
    /*...*/
}
```

If a keyword is used, it can be short, but needs to be descriptive. Maybe `keep` or `capture`?. The proposed symbol delimiters are not especially readable for me.

**Comment by dedesite** [comment]
I looked at Rust closure's example, but didn't really understand : https://doc.rust-lang.org/rust-by-example/fn/closures.html
They create closure with `||` instead of `()` eg : 
```rust
let closure_annotated = |i: i32| -> i32 { i + 1 };
```

But calling them as normal functions...
```rust
 let i = 1;
println!("closure_annotated: {}", closure_annotated(i));
```

For me, when calling a closure, you shouldn't specify the variable as a parameter, otherwise, it's just a normal function. Can someone enlighten me ?

it is difficult to choose between `||` or with a specific keyword like `use` or `with`, but I think making variable use explicit is a good idea and avoid having to share local variables with all anonymous function, which will certainly result in better C code.
Don't know how closure are implemented in other language, but bet it adds an overhead.

**Comment by Nv7-GitHub** [comment]
I think the compiler should figure out which variables are captured by checking for identifiers within the function that are not in the scope of that function but are in the upper scope. This would make it much simpler, and more like Go.

**Comment by Ben-Fields** [comment]
@dedesite The key is on the very next page of the link you gave. The first page just defines notation, and the provided example is, yes, exactly like a function. "Closures" allow you to capture variables from the context in which they are defined, and use them without passing them as parameters in future calls (they are basically passed by reference inside the closure variable). This would loosen the current scoping restrictions of V a bit in a convenient way. It looks like in Rust, these captured variables are implicit, which is why the header looks like a plain function. In V, I have argued, for example, that the captured variables should be explicit since it would follow V's established guidelines; otherwise, there would be unstated side-effects to a closure. (We  must use `mut` explicitly to indicate that regular functions have side-effects).

**Comment by dedesite** [comment]
Thanks for your reply @Ben-Fields, I realize now what was problematic for me is that Rust call closure the lambda syntax (which is also indeed a closure) and does not specified which variables are borrow from the outer scope.
The fact that several example didn't used variables from outer scope just added to the confusion ;).
